### ISSUE TITLE
function AST in a module limited to 2^16 constants

### ISSUE THREAD

**Issue by rcnlee** [JuliaLang/julia#14113]
Hello,
I am seeing memory corruption when repeatedly using ```@eval```.  Here is some example code to reproduce it.  I have it failing on both Julia 0.3 and 0.4 at around 33000 iterations, where the eval'ed function hits corruption.
Thanks!

```julia
function to_function(code::Expr)
  @eval f() = $code
  return f
end

function get_code()
  r1 = rand()
  r2 = rand()
  Expr(:comparison, r1, :<, r2)
end

using Debug
@debug function script1()
  srand(0)
  for i = 1:100000
    code = "nothing"
    try
      code = get_code()
      f = to_function(code)
      f()
    catch e
      println(e)
      @bp
    end
    println("$i: code=$(string(code))")
  end
end
```

**Comment by yuyichao** [comment]
Is `@debug` necessary to trigger the issue?

**Comment by yuyichao** [comment]
Also could you please specify what exatly do you mean by "corruption" (supposedly the program crashes with some error messages, what's the errors you see)

**Comment by rcnlee** [comment]
Thanks for the quick response.  The debug is not required to trigger the problem.  This code:
```julia
function to_function(code::Expr)
  @eval f() = $code
  return f
end

function get_code()
  r1 = rand()
  r2 = rand()
  Expr(:comparison, r1, :<, r2)
end

function script1()
  srand(0)
  for i = 1:100000
    code = "nothing"
    try
      code = get_code()
      f = to_function(code)
      f()
    catch e
      println(e)
      println(code)
      @assert false
    end
    println("$i: code=$(string(code))")
  end
end
```
results in this output:
```julia
32755: code=0.7829824199772972 < 0.9875299593717735
32756: code=0.2726528909230177 < 0.285731785216383
32757: code=0.8079519420791403 < 0.552744807360769
32758: code=0.8932454646073495 < 0.7764627836294602
32759: code=0.6841247414187677 < 0.7857370975102516
MethodError(isless,("nothing",Base))
0.6144062920627698 < 0.889133530982475
ERROR: assertion failed: false
 in script1 at C:\Users\rcnlee\.julia\v0.3\RLESCAS\src\clustering\experiments\grammar_based\grammar_test\endurance_minim
al2.jl:23
```

**Comment by mykelk** [comment]
Also fails (deterministically?) on 0.4.1.
```
32754: code=0.9381456172357221 < 0.2593768237929923
32755: code=0.7829824199772972 < 0.9875299593717735
32756: code=0.2726528909230177 < 0.285731785216383
MethodError(isless,(0.8079519420791403,Type{LabelNode}))
0.8079519420791403 < 0.552744807360769
ERROR: AssertionError: false
 in script1 at none:12

julia> versioninfo()
Julia Version 0.4.1
Commit cbe1bee* (2015-11-08 10:33 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

**Comment by rcnlee** [comment]
Here is a simpler example with the functions and debugs stripped out:
```julia
srand(0)
for i = 1:100000
  r1, r2 = rand(2)
  code = Expr(:comparison, r1, :<, r2)
  println("$i: code=$(string(code))")
  @eval f() = $code
  f()
end
```
gives this error:
```julia
32759: code=0.6841247414187677 < 0.7857370975102516
32760: code=0.6144062920627698 < 0.889133530982475
32761: code=0.6844753513581316 < 0.5679463365766453
32762: code=0.22980262610895807 < 0.9627490951848963
ERROR: `isless` has no method matching isless(::Float64, ::Type{Nothing})
 in f at endurance_minimal2.jl:12
 in anonymous at no file:13
 in include at boot.jl:245
 in include_from_node1 at loading.jl:128
while loading endurance_minimal2.jl, in expression starting on line 8
```
``isless`` should always be called with two Float64's, but at some point the arguments get corrupted to unrelated symbols.  Fails at the same number and with the same error message.

**Comment by sbromberger** [comment]
I'm seeing the same thing but at a different point / different type:

```
32754: code=0.9381456172357221 < 0.2593768237929923
32755: code=0.7829824199772972 < 0.9875299593717735
ERROR: MethodError: `isless` has no method matching isless(::Float64, ::Type{Type{LabelNode}})
Closest candidates are:
  isless(::Float64, ::Float64)
  isless(::AbstractFloat, ::AbstractFloat)
  isless(::Real, ::AbstractFloat)
  ...
 in f at ./none:5
 [inlined code] from ./none:6
 in anonymous at ./no file:4294967295
```

**Comment by yuyichao** [comment]
Just an update, I can reproduce it on master with llvm 37. The issue is pretty weird and somehow a `jl_uncompress_ast(li, jl_compress_ast(li, li->ast))` round trip changes the AST from the correct one to something that doesn't make sense...

**Comment by vtjnash** [comment]
This isn't entirely valid code due to the abuse of `@eval`. The warning is suppressed because it can be useful for iterative development in the REPL, but in general this style has very bad performance and memory issues. The net result is that you've overflowed the number of floating point constant literals allowed as code literals.

The correct formulation of this code is nearly 50_000x faster:
```
function get_code()
  r1 = rand()
  r2 = rand()
  () -> r1 < r2
end
```

The iteration count should have made this easy: 32_768 is 2^15. But there are 2 constants in each AST, so there are now just over 2^16 Floats being serialized into the module global constant table and the counter accidentally wraps around.

**Comment by rcnlee** [comment]
Thanks for looking into this.  The example is motivated by Genetic Programming work that I am doing where I am randomly generating ``Expr``'s and then transforming them into executable code.  (I am using the GrammaticalEvolution.jl package).  So unfortunately I don't think I can get away from first generating an ``Expr``.  Is there a way to transform without using ``@eval`` and/or avoid adding to the module's global constant table? e.g., reusing or cleaning it up?

**Comment by vtjnash** [comment]
If you don't mind the overhead anyways, making a new module for each function will work around this.

**Comment by rcnlee** [comment]
I'm not quite sure how to do that.  Can you give me some example code?

I'm trying to understand this a bit better.  What are the differences between these two codes?
```julia
function get_code()
  r1 = rand()
  r2 = rand()
  (r) -> r < r1 < r2
end

function get_code1()
  r1 = rand()
  r2 = rand()
  ex = Expr(:comparison, :r, :<, r1, :<, r2)
  @eval (r) -> $ex
end
```
They produce different ``code_lowered``'s.  First one is fast and doesn't fail.  But the second one is (I thought) the programmatic way of generating the first one.  Is there a different way? The first one seems to do boxing of r1 and r2, whereas the second one adds r1 and r2 to a global constant table?

Thanks!

**Comment by yuyichao** [comment]
`eval` is in global scope

**Comment by toivoh** [comment]
I think the most important difference in this case is that `get_code` is compiled once, producing code that can deal with any `r1` and `r2` values that might be supplied afterwards. `get_code1` compiles the code of the function to be returned anew for each invocation, hardcoding the current values of `r1` and `r2` into it. Since it gets invoked to compile so many different functions, it gradually consumes a lot of compiler resources. 

But if you want to do genetic programming I guess that you will have to compile lots and lots of different functions, which would be a lot more different than the ones in the example. Actually, it would probably be better if they could be interpreted instead of compiled, but I don't think there's a way to do that for functions so far.

**Comment by vtjnash** [comment]
fixed by 34d8b8e7a049b4b8bdfac85de89a5caa9180f581

**Comment by StefanKarpinski** [comment]
Can we backport that fix?
